Replace mmap with aligned weight loading

cgbur opened the pull request:
After our discussion on cache alignment in #94, I was curious to know what the performance boost would be of getting the weights properly aligned. The easiest way to do this was switch from mmap to manually loading the weights. This allows us to skip changing the file format because we can first read in the config, then the rest of the weights on whatever alignment we like. Implementing this [in my Zig fork](https://github.com/cgbur/llama2.zig/commit/23c0711308cfe52971834b3c5716e497e7d6dc3d) was a small speed win, but an even bigger one here. I chose page alignment for two reasons. 1) it was the alignment of the previous mmap data 2) it will not share a page with any other data and forevermore have the perfect alignment. 

EDIT: I have reverted the page part as I could not find a good cross platform way that did not involve some ugliness. The results for the three are in the below table, see the comments for an image of the comparison

| Metric         | Mmapped  | Aligned | Normal Aligned | Aligned vs Mmapped (%) | Normal Aligned vs Mmapped (%) |
|----------------|----------|---------|----------------|------------------------|-------------------------------|
| Average Speed  | 496.50   | 623.88  | 596.86         | 25.66%                 | 20.21%                         |
| Max Speed      | 513.95   | 646.20  | 626.06         | 25.73%                 | 21.81%                         |


Here's the comparison for the 15m parameter model:

| Metric        | Mmapped | Aligned | Percent Difference |
| ------------- | ------- | ------- | ------------------ |
| Average Speed | 496.50  | 623.88  | 25.66%             |
| Max Speed     | 513.95  | 646.20  | 25.73%             |

Here's the comparison for the 42m parameter model:

| Metric        | Mmapped | Aligned | Percent Difference |
| ------------- | ------- | ------- | ------------------ |
| Average Speed | 176.39  | 232.95  | 32.07%             |
| Max Speed     | 179.44  | 242.05  | 34.89%             |

As you can see the speedup is quite significant for a small code change. I would be curious to see the results tested on an macbook as well as on larger (7b) parameter models. 

I've also noticed this speedup mostly happens on the runfast with and without `march=native`. If you wish to reproduce my results, my make command is listed below.

![image](https://github.com/karpathy/llama2.c/assets/9002722/00b71537-3f89-4cc6-aa36-8320d8383660)
![image](https://github.com/karpathy/llama2.c/assets/9002722/ce4c977d-d9cc-4a50-a64d-991ac1c9264c)

All benchmarks ran with the following make configuration on an AMD 5900x

```make
.PHONY: runfast
runfast: run.c
	$(CC) -Ofast -o run run.c -lm -march=native

```
Fish to benchmark
```fish
 while true; ./run stories42M.bin -t 0 | rg "tok" ; end
```

cgbur committed:
Replace mmap with aligned weight loading

cgbur commented:
This is failing on windows and mac, I will need a day or two to figure this out. If anyone knows the fix please let me know. It seems my assumption that querying the page size was _not_ as cross platform as I was led to believe :upside_down_face:

kroggen commented:
This is on a MacBook Pro (Intel):

```
$ make runfast
gcc -Ofast -o run run.c -lm
$ ./run stories110M.bin 
aligned_alloc failed!
```

cgbur committed:
Use normal malloc instead of aligned

cgbur commented:
Thanks for trying @kroggen, I don't think there is a great way to to page aligned alloc without introducing more code/macros. For now I have reverted the page part and updated the PR description

kroggen commented:
Anyway, I sent you a PR

The documentation is [here](https://en.cppreference.com/w/c/memory/aligned_alloc)

For Windows:

> Also, this function is not supported in Microsoft C Runtime library because its implementation of free is unable to handle aligned allocations of any kind. Instead, MSVC provides `_aligned_malloc` (to be freed with `_aligned_free`).

So, yes, it would require macro for Windows.

kroggen commented:
There is no improvement on performance on Mac

cgbur commented:
> There is no improvement on performance on Mac

Thanks for benchmarking. I just rented a cloud instance AMD server and observed no  difference for either the page aligned or memory one. I am really curious why I am observing this large performance difference on this particular box then. I am going to mark this as a draft for now.

twobob commented:
` .\build_macro_msvc.bat`
Number of Logical Processors: 4
Using 3 processors for compilation

Compile Flags: /c /fp:fast /favor:AMD64 /Ox /Ob2 /Qpar /nologo /openmp:experimental /MP3 /MT /TC /D"COMPILER=\"MSVC\"" /I. /I"C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt" /I"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.36.32532\include" /I"C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\um" /I"C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\shared" _aligned_malloc_macro.c
_aligned_malloc_macro.c

Link Flags: /out:_aligned_malloc_macro.exe /nologo /LIBPATH:"C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.36.32532\lib\x64" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.19041.0\um\x64" /LIBPATH:"C:\Program Files (x86)\Windows Kits\10\Lib\10.0.19041.0\ucrt\x64" _aligned_malloc_macro.obj

` .\_aligned_malloc_macro.exe ../out`

> Memory for model110M.bin is 1024-byte aligned: 0000018A492F3400
> Attempting to free: ../out\model110M.bin data which is 438381596
> Memory for model110M.pt is 1024-byte aligned: 0000018A492FF400
> Attempting to free: ../out\model110M.pt data which is 438163693
> Memory for model15M.bin is 1024-byte aligned: 0000018A492F7400
> Attempting to free: ../out\model15M.bin data which is 60816028
> Memory for model15M.pt is 1024-byte aligned: 0000018A492FE400
> Attempting to free: ../out\model15M.pt data which is 60789090
> Memory for model260K.bin is 1024-byte aligned: 0000018A492F1400
> Attempting to free: ../out\model260K.bin data which is 1056540
> Memory for model260K.pt is 1024-byte aligned: 0000018A492F7400
> Attempting to free: ../out\model260K.pt data which is 1061090
> Memory for model42M.bin is 1024-byte aligned: 0000018A492F5400
> Attempting to free: ../out\model42M.bin data which is 167020572
> Memory for model42M.pt is 1024-byte aligned: 0000018A492F3400
> Attempting to free: ../out\model42M.pt data which is 166787298
> Memory for stories260K.bin is 1024-byte aligned: 0000018A492F3400
> Attempting to free: ../out\stories260K.bin data which is 1056540
> Memory for stories260K.pt is 1024-byte aligned: 0000018A492FD400
> Attempting to free: ../out\stories260K.pt data which is 1061090
> Memory for tok32k.bin is 1024-byte aligned: 0000018A49202800
> Attempting to free: ../out\tok32k.bin data which is 432717
> Memory for tok32k.model is 1024-byte aligned: 0000018A49202800
> Attempting to free: ../out\tok32k.model data which is 499723
> Memory for tok512.bin is 1024-byte aligned: 0000018A49202800
> Attempting to free: ../out\tok512.bin data which is 6227
> Memory for tok512.model is 1024-byte aligned: 0000018A49202800
> Attempting to free: ../out\tok512.model data which is 7645
> Memory for tokenizer.model is 1024-byte aligned: 0000018A49202800
> Attempting to free: ../out\tokenizer.model data which is 499723
> 
> 
> Summary:
> Successful Operations: 15
> Failed Operations: 0
> 

https://github.com/twobob/llama2.c/blob/master/_aligned_malloc_macro.c

maybe? does that help?

I just had it read over a few relevant files?

kroggen commented:
Check this [PR](https://github.com/kroggen/llama2.c/pull/2)

I attempted to create it on your fork but it redirected to mine
